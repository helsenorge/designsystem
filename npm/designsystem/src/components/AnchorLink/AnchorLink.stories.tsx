import React from 'react';

import { action } from '@storybook/addon-actions';
import { Meta, StoryObj } from '@storybook/react';

import AnchorLink from './AnchorLink';

const meta = {
  title: '@helsenorge∕designsystem-react/Components/AnchorLink',
  component: AnchorLink,
  parameters: {
    docs: {
      description: {
        component:
          'AnchorLink kan bli brukt som et komponent for \\<a\\> tag eller \\<button\\> tag som skal ha en anchorlink styling.<br>.anchorlink-wrapper kan også bli brukt når det ikke er kontroll over markup som kommer inn.',
      },
    },
  },
  args: {
    children: 'Anchorlink tekst',
    href: '/test',
  },
  argTypes: {
    children: {
      control: 'text',
    },
    href: {
      control: 'text',
    },
  },
} satisfies Meta<typeof AnchorLink>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Default: Story = {
  render: args => (
    <>
      <AnchorLink {...args} target={'_self'}>
        {args.children}
      </AnchorLink>
      <br />
      <br />
      <AnchorLink {...args} target={'_self'}>
        {
          'Eiusmod veniam reprehenderit dolore magna tempor dolor reprehenderit reprehenderit ullamco sit in nulla qui. (Lang tekst - Skal wrappe).'
        }
      </AnchorLink>
    </>
  ),
};

export const External: Story = {
  render: args => (
    <>
      <AnchorLink {...args} target={'_blank'}>
        {args.children}
      </AnchorLink>
      <br />
      <br />
      <AnchorLink {...args} target={'_blank'}>
        {
          'Eiusmod veniam reprehenderit dolore magna tempor dolor reprehenderit reprehenderit ullamco sit in nulla qui. (Lang tekst - Skal wrappe).'
        }
      </AnchorLink>
    </>
  ),
};

export const AsButton: Story = {
  render: args => (
    <p style={{ fontSize: '1.25rem' }}>
      Dette er først en{' '}
      <AnchorLink htmlMarkup={'a'} onClick={action('AnchorLink clicked!')} {...args}>
        vanlig lenke i løpende tekst
      </AnchorLink>{' '}
      og nå kommer en
      <AnchorLink htmlMarkup={'button'} onClick={action('AnchorLink clicked!')} {...args}>
        button-lenke i løpende tekst som går over flere linjer
      </AnchorLink>{' '}
      og til slutt en{' '}
      <AnchorLink htmlMarkup={'button'} onClick={action('AnchorLink clicked!')} {...args}>
        kort
      </AnchorLink>{' '}
      button-lenke
    </p>
  ),
};
