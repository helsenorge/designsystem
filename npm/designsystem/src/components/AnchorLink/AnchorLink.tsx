import React from 'react';
import Icon, { IconSize } from '../Icons';
import ArrowUpRight from '../Icons/ArrowUpRight';
import { getColor } from '../../theme/currys';
import { useHover } from '../../hooks/useHover';

import AnchorLinkStyles from './styles.module.scss';
import classNames from 'classnames';

export type AnchorLinkTargets = '_self' | '_blank' | '_parent';

interface AnchorLinkProps {
  /** Sets the content of the <a> tag */
  children: React.ReactNode;
  /** URL to link to */
  href: string;
  /** Gives a unique id to the anchor-link :) */
  id?: string;
  /** Adds custom classes to the element. */
  className?: string;
  /** Sets the target type of the <a> tag. _blank adds an arrow icon at the end of the link */
  target?: AnchorLinkTargets;
}

const AnchorLink = React.forwardRef(function AnchorLinkForwardedRef(props: AnchorLinkProps, ref: React.ForwardedRef<HTMLAnchorElement>) {
  const { id, href, children, className, target = '_self' } = props;
  const external = target === '_blank';
  const { hoverRef, isHovered } = useHover<HTMLAnchorElement>(ref as React.RefObject<HTMLAnchorElement>, undefined, false);
  const anchorClasses = classNames(AnchorLinkStyles.anchorlink, className ?? '');

  return (
    <a id={id} href={href} className={anchorClasses} target={target} ref={hoverRef}>
      {children}
      {external && (
        <Icon
          className={AnchorLinkStyles.anchorlink__icon}
          svgIcon={ArrowUpRight}
          color={getColor('blueberry', 600)}
          hoverColor={getColor('blueberry', 700)}
          size={IconSize.XSmall}
          isHovered={isHovered}
        />
      )}
    </a>
  );
});

export default AnchorLink;
