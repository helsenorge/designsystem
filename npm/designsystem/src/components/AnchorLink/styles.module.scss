@use 'sass:map';
@import '../../scss/spacers';
@import '../../scss/breakpoints';
@import '../../scss/palette';

$underline-width: 0.0625rem;
$focus-width: 0.15rem;

@mixin anchorlink-hover {
  cursor: pointer;
  color: $blueberry700;
  background-color: $blueberry60010;
  text-decoration-color: $blueberry600;
}

@mixin anchorlink-focus {
  outline: none;
  border: $focus-width solid $black;
}

@mixin anchorlink {
  display: inline;
  box-decoration-break: clone;
  overflow-wrap: break-word;
  word-break: break-word;
  text-align: left;
  color: $blueberry600;
  border: $focus-width solid transparent;
  text-decoration: underline;
  text-underline-offset: 0.16rem;
  text-decoration-color: $blueberry400;
  text-decoration-thickness: $underline-width;
  padding: 0.1rem;

  &:hover {
    @include anchorlink-hover;
  }

  &:focus {
    @include anchorlink-focus;
  }

  &__icon {
    display: inline-block;
    vertical-align: bottom;
    margin-right: -0.4rem;
    margin-top: -2rem;
    margin-bottom: -0.3rem;
  }
}

@mixin anchorlink-icon {
  &[target='_blank'] {
    &::after {
      content: url("data:image/svg+xml;utf8,<svg xmlns='http://www.w3.org/2000/svg'  width='38px' height='38px' viewBox='0 0 48 48' fill='%2308667c'><polygon points='33.305,14.659,33.305,29.205,32.005,29.205,32.005,16.948,15.804,33.149,14.885,32.229,31.155,15.959,18.761,15.96,18.761,14.659' /></svg>");
      display: inline-block;
      vertical-align: bottom;
      margin-right: -0.4rem;
      margin-top: -2rem;
      margin-bottom: -0.8rem;
    }
  }
}

.anchorlink {
  @include anchorlink;
}

.anchorlink-wrapper {
  all: unset;

  &:hover > span {
    @include anchorlink-hover;
  }

  &:focus > span {
    @include anchorlink-focus;
  }
}
