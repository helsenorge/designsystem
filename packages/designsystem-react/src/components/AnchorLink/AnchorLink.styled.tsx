import styled from 'styled-components';
import {typography} from '../../theme/typography';
import {getColor} from '../../theme/currys';

interface StyledAnchorLinkProps {
  external: boolean;
}

const StyledAnchorLink = styled('a')<StyledAnchorLinkProps>`
  display: inline;
  overflow-wrap: break-word;
  word-wrap: break-word;
  text-align: left;
  padding-bottom: 0.15rem;
  text-decoration: none;
  ${typography['anchorlink']};
  color: ${getColor('blueberry', 600)};
  border-bottom: 1px solid ${getColor('neutral', 300)};
  .hnds-style-icon {
    vertical-align: bottom;
    margin-bottom: -0.25rem;
  }
  &:focus {
    outline: none;
    background-color: rgba(8, 102, 124, 0.05); /* Dette er blueberry 700 med 5% opacity*/
    border-color: ${getColor('neutral', 300)};
  }
  &:hover {
    cursor: pointer;
    color: ${getColor('blueberry', 700)};
    border-color: ${getColor('blueberry', 600)};
  }
`;

export default StyledAnchorLink;
