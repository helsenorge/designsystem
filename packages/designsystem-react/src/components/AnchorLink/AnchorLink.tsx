import React from 'react';
import Icon from '../Icons';
import ArrowUpRight from '../Icons/ArrowUpRight';
import StyledAnchorLink from './AnchorLink.styled';
import {getColor} from '../../theme/currys';
import {useHover} from '../../hooks/useHover';

export type AnchorLinkTargets = '_self' | '_blank' | '_parent';

interface AnchorLinkProps {
  children: React.ReactNode;
  href: string;
  id?: string;
  className?: string;
  target?: AnchorLinkTargets;
}

const AnchorLink = React.forwardRef((props: AnchorLinkProps, ref: any) => {
  const {id, href, children, className, target = '_self'} = props;
  const external = target === '_blank';
  const {hoverRef, isHovered} = useHover<HTMLAnchorElement>(ref);
  return (
    <StyledAnchorLink id={id} href={href} className={className} target={target} external={external} ref={hoverRef}>
      {children}
      {external && (
        <Icon
          svgIcon={ArrowUpRight}
          color={getColor('blueberry', 600)}
          hoverColor={getColor('blueberry', 700)}
          size={38}
          isHovered={isHovered}
        />
      )}
    </StyledAnchorLink>
  );
});

export default AnchorLink;
