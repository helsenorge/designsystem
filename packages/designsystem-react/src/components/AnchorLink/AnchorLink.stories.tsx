import React from 'react';
import {withKnobs, text} from '@storybook/addon-knobs';
import {storiesOf} from '@storybook/react';
import AnchorLink from './AnchorLink';
import {withA11y} from '@storybook/addon-a11y';

import './../../scss/typography.module.scss';

const stories = storiesOf('AnchorLink', module);
stories.addDecorator(withKnobs);
stories.addDecorator(withA11y);

stories.add('Default', () => (
  <div style={{width: '20rem'}}>
    <AnchorLink href={text('Href', '/test')} target={'_self'}>
      {text('Text', 'link text')}
    </AnchorLink>
    <br />
    <br />
    <AnchorLink href={text('Href', '/test')} target={'_self'}>
      {'I am a super long link (I am supposed to wrap!)'}
    </AnchorLink>
  </div>
));

stories.add('External', () => (
  <div style={{width: '20rem'}}>
    <AnchorLink href={text('Href', '/test')} target={'_blank'}>
      {text('Text', 'link text')}
    </AnchorLink>
    <br />
    <br />
    <AnchorLink href={text('Href', '/test')} target={'_blank'}>
      {'I am a super long link (it is supposed to wrap!)'}
    </AnchorLink>
  </div>
));

stories.add('anchorlink-wrapper', () => (
  <div className={'anchorlink-wrapper'} style={{width: '20rem'}}>
    <p>
      {'Lorem ipsum dolor sit amet sonsectur '}
      {' amet consectur. Lorem a '}
      <a href={'#test'} target={'_blank'}>
        {'click me'}
      </a>
      {'Lorem ipsum dolor sit amet sonsectur '}
    </p>
  </div>
));
