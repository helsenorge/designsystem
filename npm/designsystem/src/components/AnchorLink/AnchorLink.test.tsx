import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import AnchorLink from './AnchorLink';

describe('Gitt at AnchorLink skal rendres', (): void => {
  describe('Gitt at AnchorLink skal vises vanlig', (): void => {
    test('Så rendres AnchorLink riktig', (): void => {
      const { container } = render(<AnchorLink target={'_self'}>{'AnchorLink'}</AnchorLink>);
      expect(container).toMatchSnapshot();

      const svgElement = screen.queryByRole('img', { hidden: true });
      expect(svgElement).not.toBeTruthy();
    });
  });

  describe('Gitt at AnchorLink skal vises med className og id', (): void => {
    test('Så rendres AnchorLink riktig', (): void => {
      const { container } = render(
        <AnchorLink target={'_self'} id={'custom-id'} className={'custom-classname'}>
          {'AnchorLink'}
        </AnchorLink>
      );
      expect(container).toMatchSnapshot();
    });
  });

  describe('Gitt at AnchorLink skal vises som external', (): void => {
    test('Så rendres AnchorLink riktig', (): void => {
      const { container } = render(<AnchorLink target={'_blank'}>{'AnchorLink external'}</AnchorLink>);
      expect(container).toMatchSnapshot();
    });

    test('Så har AnchorLink riktig rel-attributt', (): void => {
      render(
        <AnchorLink testId="target-test" target={'_blank'}>
          {'AnchorLink external'}
        </AnchorLink>
      );
      const component = screen.getByTestId('target-test');
      expect(component).toBeVisible();
      expect(component).toHaveAttribute('rel', 'noopener noreferrer');
    });

    test('Så har AnchorLink ArrowUpRight svg', (): void => {
      render(<AnchorLink target={'_blank'}>{'AnchorLink external'}</AnchorLink>);

      const svgElement = screen.getByRole('presentation', { hidden: true });
      expect(svgElement).toBeTruthy();

      // eslint-disable-next-line testing-library/no-node-access
      const pathElement = svgElement.firstChild;
      expect(pathElement).toHaveAttribute('d', 'M18.76 14.426v1.768h11.83L14.719 32.065l1.251 1.25 15.801-15.801v11.691h1.768V14.426z');
    });
  });

  describe('Når testId-prop er satt', (): void => {
    test('Så kan komponenten finnes ved hjelp av testId', (): void => {
      render(
        <AnchorLink testId="bare-tester" href="/">
          {'AnchorLink'}
        </AnchorLink>
      );

      const component = screen.getByTestId('bare-tester');
      expect(component).toBeVisible();
    });
  });

  describe('Når htmlMarkup er button', () => {
    test('Skal kalles onClick-handler når man klikker på knappen', async () => {
      const onClickMock = vi.fn();
      render(
        <AnchorLink htmlMarkup="button" onClick={onClickMock} href="#">
          {'Lenketekst'}
        </AnchorLink>
      );

      const link = screen.getByRole('button', { name: 'Lenketekst' });
      await userEvent.click(link);
      expect(onClickMock).toHaveBeenCalledTimes(1);
    });
  });

  describe('Når htmlMarkup er a', () => {
    test('Skal kalles onClick-handler når man klikker på knappen', async () => {
      const onClickMock = vi.fn();
      render(
        <AnchorLink htmlMarkup="a" onClick={onClickMock} href="#">
          {'Lenketekst'}
        </AnchorLink>
      );

      const link = screen.getByRole('link', { name: 'Lenketekst' });
      await userEvent.click(link);
      expect(onClickMock).toHaveBeenCalledTimes(1);
    });
  });
});
