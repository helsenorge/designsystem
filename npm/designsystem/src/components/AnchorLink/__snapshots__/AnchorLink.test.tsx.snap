// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Gitt at AnchorLink skal rendres > Gitt at AnchorLink skal vises med className og id > Så rendres AnchorLink riktig 1`] = `
<div>
  <a
    class="anchorlink custom-classname"
    data-analyticsid="anchor-link"
    id="custom-id"
    target="_self"
  >
    AnchorLink
  </a>
</div>
`;

exports[`Gitt at AnchorLink skal rendres > Gitt at AnchorLink skal vises som external > Så rendres AnchorLink riktig 1`] = `
<div>
  <a
    class="anchorlink"
    data-analyticsid="anchor-link"
    rel="noopener noreferrer"
    target="_blank"
  >
    AnchorLink external
    <svg
      aria-hidden="true"
      class="hnds-style-icon anchorlink__icon"
      color="#126F87"
      data-analyticsid="icon"
      fill="#126F87"
      focusable="false"
      height="38"
      role="presentation"
      style="min-width: 38px; min-height: 38px;"
      viewBox="0 0 48 48"
      width="38"
    >
      <path
        d="M18.76 14.426v1.768h11.83L14.719 32.065l1.251 1.25 15.801-15.801v11.691h1.768V14.426z"
      />
    </svg>
  </a>
</div>
`;

exports[`Gitt at AnchorLink skal rendres > Gitt at AnchorLink skal vises vanlig > Så rendres AnchorLink riktig 1`] = `
<div>
  <a
    class="anchorlink"
    data-analyticsid="anchor-link"
    target="_self"
  >
    AnchorLink
  </a>
</div>
`;
